packages/react-transform/src/index.ts

```diff
@@ -0,0 +1,369 @@
+import {
+	types as BabelTypes,
+	template as BabelTemplate,
+	PluginObj,
+	PluginPass,
+	NodePath,
+	template,
+} from "@babel/core";
+import { isModule, addNamed, addNamespace } from "@babel/helper-module-imports";
+
+// TODO:
+// - how to trigger rerenders on attributes change if transform never sees
+//   `.value`?
+
+interface PluginArgs {
+	types: typeof BabelTypes;
+	template: typeof BabelTemplate;
+}
+
+const optOutCommentIdentifier = /(^|\s)@noTrackSignals(\s|$)/;
+const optInCommentIdentifier = /(^|\s)@trackSignals(\s|$)/;
+const dataNamespace = "@preact/signals-react-transform";
+const importSource = "@preact/signals-react/runtime";
+const importName = "useSignals";
+const getHookIdentifier = "getHookIdentifier";
+const maybeUsesSignal = "maybeUsesSignal";
+const containsJSX = "containsJSX";
+const alreadyTransformed = "alreadyTransformed";
+
+const get = (pass: PluginPass, name: any) =>
+	pass.get(`${dataNamespace}/${name}`);
+const set = (pass: PluginPass, name: string, v: any) =>
+	pass.set(`${dataNamespace}/${name}`, v);
+
+interface DataContainer {
+	getData(name: string): any;
+	setData(name: string, value: any): void;
+}
+const setData = (node: DataContainer, name: string, value: any) =>
+	node.setData(`${dataNamespace}/${name}`, value);
+const getData = (node: DataContainer, name: string) =>
+	node.getData(`${dataNamespace}/${name}`);
+
+function setOnFunctionScope(path: NodePath, key: string, value: any) {
+	const functionScope = path.scope.getFunctionParent();
+	if (functionScope) {
+		setData(functionScope, key, value);
+	}
+}
+
+type FunctionLike =
+	| BabelTypes.ArrowFunctionExpression
+	| BabelTypes.FunctionExpression
+	| BabelTypes.FunctionDeclaration;
+
+function testFunctionName<T extends FunctionLike>(predicate: (name: string | null) => boolean): (path: NodePath<T>) => boolean {
+	return (path: NodePath<T>) => {
+		if (
+			path.node.type === "ArrowFunctionExpression" ||
+			path.node.type === "FunctionExpression"
+		) {
+			return (
+				path.parentPath.node.type === "VariableDeclarator" &&
+				path.parentPath.node.id.type === "Identifier" &&
+				predicate(path.parentPath.node.id.name)
+			);
+		} else if (path.node.type === "FunctionDeclaration") {
+			return predicate(path.node.id?.name ?? null);
+		} else {
+			return false;
+		}
+	};
+}
+
+const fnNameStartsWithCapital = testFunctionName(name => name?.match(/^[A-Z]/) !== null);
+const fnNameStartsWithUse = testFunctionName(name => name?.match(/^use[A-Z]/) !== null);
+
+function hasLeadingComment(path: NodePath, comment: RegExp): boolean {
+	const comments = path.node.leadingComments;
+	return comments?.some(c => c.value.match(comment) !== null) ?? false;
+}
+
+function hasLeadingOptInComment(path: NodePath) {
+	return hasLeadingComment(path, optInCommentIdentifier);
+}
+
+function hasLeadingOptOutComment(path: NodePath) {
+	return hasLeadingComment(path, optOutCommentIdentifier);
+}
+
+function isOptedIntoSignalTracking(path: NodePath | null): boolean {
```

Comment:
Here and in `isOptedOutOfSignalTracking`, I'm climbing up the AST to find the most logical place for a comment to live. For example supporting a use case such as:
```js
/** @trackSignals */
export default function useCustomHook() {}
```